pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -148,56 +148,166 @@ func NewFlowExporter(connStore *connections.ConntrackConnectionStore, records *f
+					if exp.process != nil {
+						exp.process.CloseConnToCollector()
+						exp.process = nil
+					}
+					// Initializing flow exporter fails, will retry in next cycle.
+					expireTimer.Reset(exp.activeFlowTimeout)
+					continue
+				}
+			}
+			// Pop the flow record or deny connection item from expire priority queue
+			// in the exporting process and send the data set.
+			if err := exp.sendFlowRecords(); err != nil {
+				klog.Errorf("Error when sending expired flow records: %v", err)
+				// If there is an error when sending flow records because of intermittent
+				// connectivity, we reset the connection to IPFIX collector and retry
+				// in the next export cycle to reinitialize the connection and send flow records.
+				exp.process.CloseConnToCollector()
+				exp.process = nil
+				expireTimer.Reset(exp.activeFlowTimeout)
+				continue
+			}
+
+			exp.expirePriorityQueue.Mutex.Lock()
+			// Get the new expiry and reset the timer.
+			expireTimer.Reset(exp.GetExpiryFromExpirePriorityQueue())
+			exp.expirePriorityQueue.Mutex.Unlock()
 		}
 	}
-	// Send flow records to IPFIX collector.
-	err := exp.sendFlowRecords()
-	if err != nil {
-		klog.Errorf("Error when sending flow records: %v", err)
-		// If there is an error when sending flow records because of intermittent connectivity, we reset the connection
-		// to IPFIX collector and retry in the next export cycle to reinitialize the connection and send flow records.
-		exp.process.CloseConnToCollector()
-		exp.process = nil
-		return
+}
+
+// GetExpiryFromExpirePriorityQueue returns the earliest timestamp (active expiry
+// or idle expiry) from expire priority queue.
+func (exp *flowExporter) GetExpiryFromExpirePriorityQueue() time.Duration {
+	currTime := time.Now()
+	if exp.expirePriorityQueue.Len() > 0 {
+		// Get the minExpireTime of the top item in ExpirePriorityQueue.
+		expiryDuration := MinExpiryTime + exp.expirePriorityQueue.MinExpireTime(0).Sub(currTime)
+		if expiryDuration < 0 {
+			return MinExpiryTime
+		}
+		return expiryDuration
+	}
+	if exp.activeFlowTimeout < exp.idleFlowTimeout {
+		return exp.activeFlowTimeout
 	}
-	klog.V(2).Infof("Successfully exported IPFIX flow records")
+	return exp.idleFlowTimeout
+}
+
+func (exp *flowExporter) sendFlowRecords() error {
+	exp.expirePriorityQueue.Mutex.Lock()
+	defer exp.expirePriorityQueue.Mutex.Unlock()
+
+	if exp.expirePriorityQueue.Len() == 0 {
+		return nil
+	}
+	currTime := time.Now()
+	for exp.expirePriorityQueue.Len() > 0 {
+		topItem := exp.expirePriorityQueue.Peek()
+		// If the top item is not timeout, then we do not have to check the following items.
+		if topItem.ActiveExpireTime.After(currTime) && topItem.IdleExpireTime.After(currTime) {
+			break
+		}
+		pqItem := heap.Pop(exp.expirePriorityQueue).(*flowexporter.ItemToExpire)
+		// The pqItem contains a flow record.
+		if !pqItem.IsDenyConn {
+			if pqItem.IsActive {
+				// Send a flow record if it is active and time out.
```

Comment:
Have a lot of comments containing words like "flow record", not rephrased yet. Not sure whether to keep them as they are or replace "flow record" with "pqItem" or "connection".